gpioioctl:Implement ioctl access to Linux GPIO chips/lines.

gsexton opened the pull request:
Fixes host issue #39

First, let me start by saying I tend to work in lots of languages, and my fluency in golang is not really high. If you see non-idiomatic things you can't live with, just let me know. Similarly, I know you're the maintainer, and it's your call in implementations.

This is a pretty big add. If you feel like a call would be beneficial, I can do a voice call.

## Terminology

I'm using the term "Line", and not "Pin". Line does implement conn/v3/gpio.PinIn, PinOut, and PinIO. This is kind of a difference. Line is an attribute of the chip, while pin is an attribute of the board. It seems to me that we want the code to work across boards, regardless of how a board physically presents a GPIO chipset's line. So a pin is a physical element that
the chip's line is connected to. I can see potentially a GPIO line that's not exposed as a physical pin on a board. As an aside, the ioctl()s use the Line terminology. However, if you think that's unnecessarily confusing, I can change it.

## Tests
Refer to Tests.md

## Interfaces

### consumer

In the ioctl(), there's a "consumer" field. The kernel docs say:

```
    a functional name for the consumer of this GPIO line as set by whatever is using it,
    will be empty if there is no current user but may also be empty if the consumer doesn’t
    set this up
```

I'm writing the program_name@pid into this field. I can see other programs use it too. For example, I see PPS bound to a GPIO Chip on my Pi. You might consider adding 

func Consumer() string

to the gpio.Pin interface for the next revision.

### PinFunc

I really don't understand the PinFunc interface, so I didn't implement it. I looked it over, and from what I could tell, I didn't like the idea, but perhaps I didn't understand it. If you can give me a better idea of the intent, I'll give it a go.

### PWM

PWM isn't really part of the GPIO ioctl User Space API. The kernel docs say:
```
Do NOT abuse userspace APIs to control hardware that has proper kernel drivers.
There may already be a driver for your use case, and an existing kernel driver
is sure to provide a superior solution to bitbashing from userspace.
```
I've stubbed in a PWM method to satisfy PinOut, but it should probably be deprecated from the PinOut interface. Here's the what the doc says:
```
pwm-gpio: drivers/pwm/pwm-gpio.c is used to toggle a GPIO with a high resolution
timer producing a PWM waveform on the GPIO line, as well as Linux high resolution
timers can do.
```
## MultiPin IO
Multi-Pin IO is handled by the new construct LineSet. The ioctl() interface makes this pretty straight forward. Single line input/output are identical to multi-line. Other than configuration, no clever tricks are required. You just call Read() and it
returns the pin values as a uint64 bitmap. Essentially, a LineSet provides methods for reading or writing all lines in one go. LineSetLine provides convenience methods to allow writing individual pins and implements PinIn, PinOut, and PinIO.  Additionally, functions of pins can be different within the set. Some can be input, and others output.

A common use case for a LineSet might be doing a HD44780 LCD display which has 4 or 8 lines + a strobe line, and a backlight line.

maruel commented on gpioioctl/Testing.md:
```diff
@@ -0,0 +1,73 @@
+# Testing
```
This should go in https://github.com/periph/cmd/tree/main/periph-smoketest/gpiosmoketest

gsexton replied:
I've done the restructure. The sysfs stuff is still conflicting. Do you understand it?

maruel commented:
I haven't looked at the code yet but answering a few questions for sake of speed:
- PinFunc is mostly relevant for MCU provided hardware pins, so it's okay to ignore for now.
- PWM: sounds good. It's okay to just return an error.
- Line vs Pin and multi-pins: it's something I wanted to create a conn/v4 for but lost steam sadly as the refactor became big; I want to make all stateful functions like PWM() to accept a context.Context.

maruel commented on gpioioctl/basic_test.go:
```diff
@@ -0,0 +1,120 @@
+	if len(chip.Name()) == 0 {
+		t.Error("chip.Name() is 0 length")
+	}
+	if len(chip.Path())==0 {
```
You may want to configure gofmt -w -s and/or goimports -w automatically in your IDE. The github checks won't pass otherwise.

maruel commented on gpioioctl/basic_test.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+func TestChips(t *testing.T) {
+	if len(Chips) <= 0 {
```
Can this test pass on macOS and Windows? Options could be to mock or move this kind of testing to https://github.com/periph/cmd/pull/7.

gsexton replied:
Can we do this with build tags and conditional compilation? Do you specify any --tags args in the go test command?

maruel replied:
Even so, the test assumes the host will expose chips on linux. This will not be the case on VMs.

gsexton replied:
OK, I've added a mock to make this work regardless if init() finds any gpio chips.

maruel commented on gpioioctl/basic_test.go:
```diff
@@ -0,0 +1,120 @@
+	"periph.io/x/conn/v3/gpio/gpioreg"
+)
+
+var _test_line *GPIOLine
```
FYI, go naming convention would be testLine.

maruel commented on gpioioctl/basic_test.go:
```diff
@@ -0,0 +1,120 @@
+var _test_line *GPIOLine
+
+func init() {
+	_, _ = driverreg.Init()
```
Would you mind checking for errors here? Otherwise it could create hard-to-diagnose failures.

maruel replied:
Did you forget to push a new version? It's okay to just panic.

gsexton replied:
Sorry, I somehow didn't commit this change.

maruel commented on gpioioctl/example_test.go:
```diff
@@ -0,0 +1,103 @@
+	}
+	_ = led.Out(true)
+
+	testRotary(chip, "GPIO20", "GPIO21", "GPIO19")
```
Caveat: Maybe, I'm not sure.
You may want to inline the function to make it look better. You can confirm that it looks good:
```
go install golang.org/x/pkgsite/cmd/pkgsite@latest
pkgsite
```
then navigate to http://localhost:8080/periph.io/x/host/v3

gsexton replied:
I've fixed this and it shows in the format as expected.

maruel commented on sysfs/sysfssmoketest/benchmark.go:
```diff
@@ -1,69 +0,0 @@
-// Copyright 2017 The Periph Authors. All rights reserved.
```
sysfs/sysfssmoketest/benchmark.go is a remnant where I failed (forgot!) to move it to cmd. I'm sorry about that. I'd prefer to do this as a separate PR since this PR is already fairly big.

gsexton replied:
restored.

maruel commented on gpioioctl/lineset.go:
```diff
@@ -0,0 +1,406 @@
+	return lsl.pull
+}
+
+// DefaultPull - return gpio.PullNoChange. Reviewing the GPIO v2 Kernel
```
Suggestion:

// DefaultPull return gpio.PullNoChange.
//
// The GPIO v2 Kernel doesn't support this.

maruel reviewed (commented):
Awesome. Don't be afraid by the number of comments, it's a chunk change. Thanks again.

gsexton commented on gpioioctl/gpio.go:
```diff
@@ -0,0 +1,603 @@
+
+type Label string
+
+var DirectionLabels = []Label{"NotSet", "Input", "Output"}
```
I didn't originally, but the smoke test references the values, so it's an artifact of moving functional tests out of the package.

gsexton commented on gpioioctl/ioctl.go:
```diff
@@ -0,0 +1,254 @@
+}
+
+func (ci *gpiochip_info) String() string {
+	return fmt.Sprintf("{\"Name\": \"%s\", \"Label\": \"%s\", \"Lines\": %d}", string(ci.name[:]), string(ci.label[:]), ci.lines)
```
I've escaped the JSON. I tend to do a lot of back-end work where debuggers just aren't a thing. Debugging information is what I can get in the logs, and so I tend to be pretty complete. From a support standpoint, having all of this dump in a legible format is just super-handy. 

OTOH, there's no reasonable use case for the String() to return end user displayed data here. In this case, the end users are developers. I thought about using yaml, but had already written the JSON.

maruel commented on gpioioctl/basic_test.go:
```diff
@@ -0,0 +1,120 @@
+package gpioioctl
```
The usual Go style is:

```
// Copyright 2024 The Periph Authors. All rights reserved.
// Use of this source code is governed under the Apache License, Version 2.0
// that can be found in the LICENSE file.

// Basic tests. More complete test is contained in the
// periph.io/x/cmd/periph-smoketest/gpiosmoketest
// folder.

package gpioioctl

import (
```

Note that cmd is not versioned, since it's a set of executables, not an API.

maruel commented on gpioioctl/gpio.go:
```diff
@@ -0,0 +1,617 @@
+	}
+}
+
+func escapeJSONString(str string) string {
```
FYI, the better implementation is to use:
```
d, _ := json.Marshal(str)
return string(d)
```

https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/encoding/json/encode.go;l=160 ultimately leads to https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/encoding/json/encode.go;l=967 which has many subtleties.

That sad, I don't think you need this function given the other comment.

maruel reviewed (commented):
Notes about JSON.
I changed the GitHub settings so the actions should start immediately upon push, so you don't need to wait for me to click the Approve button.

gsexton commented on gpioioctl/gpio.go:
```diff
@@ -0,0 +1,617 @@
+// String returns information about the line in JSON format.
+func (line *GPIOLine) String() string {
+
+	return fmt.Sprintf("{\"Line\": %d, \"Name\": \"%s\", \"Consumer\": \"%s\", \"Direction\": \"%s\", \"Pull\": \"%s\", \"Edges\": \"%s\"}",
```
Implemented

maruel commented on gpioioctl/ioctl.go:
```diff
@@ -0,0 +1,254 @@
+			}
+		}
+	}
+	return fmt.Sprintf("{\"Name\": \"%s\", \"Consumer\": \"%s\", \"Offset\": %d, \"# Attrs\": %d, \"Flags\": \"%s\" 0x%x, \"Attributes\": \"%s\"}",
```
Did you mean to push a new version?

gsexton replied:
Can you help me out with what I'm doing wrong? I look at the complete file for the PR, and it doesn't show the lines you're quoting.

maruel replied:
Here's what I see with the URL you pasted:
![Capture d’écran du 2024-09-09 13-25-27](https://github.com/user-attachments/assets/784abebf-1fd3-46a9-8fa0-edf952a7084e)

Here's what I'm commenting on:
![Capture d’écran du 2024-09-09 13-27-41](https://github.com/user-attachments/assets/76e756fe-7c11-4bdb-af35-d5631a82f6d7)

maruel commented:
I'll merge once the checks are green.

gsexton commented:
@maruel So i'm pretty sure the reason that netlink was not loading in host init() is because the pipeline blocked infinitely trying receive the message from the netlink socket. I got creative and added a 2s timeout to the onewire.Init(). That seems to stop the pipeline from timing out. It now passes the Linux pipeline build. I tested it on a pi with a couple of DS18B20 units attached and it worked as expected.

I've fixed all of the lint issues in gpioioctl, sysfs, and netlink, which is code that I've touched.

The rest is all gosec G115 possible overflow errors. I don't think they're false positives. I'd recommend disabling that check in the pipeline.

maruel commented on host_linux.go:
```diff
@@ -5,6 +5,8 @@
-	// Make sure sysfs drivers are registered.
+	// Make sure required drivers are registered.
+	_ "periph.io/x/host/v3/gpioioctl"
+	_ "periph.io/x/host/v3/netlink"
```
Let's do one change at a time. This change is already very large. Let's do netlink modification in a separate PR. Thanks.

gsexton replied:
OK, I've reverted those files. I'll re-open the issue I had earlier and make a new PR for it.

gsexton commented:
@maruel Were you waiting on me for something more? I think I've got everything you asked for completed.

maruel commented:
Yes sorry I just needed to silence gosec, done in https://github.com/periph/host/commit/1e513a0cdf6a479e05b38984a467787868c7d006.

maruel commented:
Thanks a lot! It's really appreciated.
